File: backend/src/serverless/microservices/python/crowd-backend/crowd/backend/models/integration.py
from sqlalchemy import Column, String, DateTime, Integer, Text, ForeignKey
from .base import Base
from sqlalchemy.orm import relationship, validates
from sqlalchemy.dialects.postgresql import JSONB


class Integration(Base):
    """
    Integration model

    Args:
        Base (Base): Parent class
    """

    __tablename__ = "integrations"  # Table name in database

    id = Column(String, primary_key=True)
    platform = Column(Text)
    status = Column(Text)
    limitCount = Column(Integer)
    limitLastResetAt = Column(DateTime)
    token = Column(Text)
    refreshToken = Column(Text)
    settings = Column(JSONB)
    integrationIdentifier = Column(Text)
    importHash = Column(String, nullable=True)
    createdAt = Column(DateTime)
    updatedAt = Column(DateTime)
    deletedAt = Column(DateTime)

    tenantId = Column(String, ForeignKey("tenants.id"), nullable=False)
    # parentTenant = relationship("Tenant", back_populates="integrations")

    createdById = Column(String, ForeignKey("users.id"))
    updatedById = Column(String, ForeignKey("users.id"))

    parentUser = relationship("User", foreign_keys=[createdById])
    updateParentUser = relationship("User", foreign_keys=[updatedById])

    @validates("importHash")
    def validate_importHash(self, key, value):
        if value is not None:
            assert len(value) <= 255
        return value

File: services/apps/webhooks_api/src/models/integration.ts
import { DataTypes } from 'sequelize'

export default (sequelize) => {
  const integration = sequelize.define(
    'integration',
    {
      id: {
        type: DataTypes.UUID,
        defaultValue: DataTypes.UUIDV4,
        primaryKey: true,
      },
      platform: {
        type: DataTypes.TEXT,
      },
      status: {
        type: DataTypes.TEXT,
      },
      limitCount: {
        type: DataTypes.INTEGER,
      },
      limitLastResetAt: {
        type: DataTypes.DATE,
      },
      token: {
        type: DataTypes.TEXT,
      },
      refreshToken: {
        type: DataTypes.TEXT,
      },
      settings: {
        type: DataTypes.JSONB,
        defaultValue: {},
      },
      integrationIdentifier: {
        type: DataTypes.TEXT,
      },
      importHash: {
        type: DataTypes.STRING(255),
        allowNull: true,
        validate: {
          len: [0, 255],
        },
      },
      emailSentAt: {
        type: DataTypes.DATE,
      },
    },
    {
      indexes: [
        {
          unique: true,
          fields: ['importHash', 'tenantId'],
          where: {
            deletedAt: null,
          },
        },
        {
          unique: false,
          fields: ['integrationIdentifier'],
        },
      ],
      timestamps: true,
      paranoid: true,
    },
  )

  integration.associate = (models) => {
    models.integration.belongsTo(models.tenant, {
      as: 'tenant',
      foreignKey: {
        allowNull: false,
      },
    })

    models.integration.belongsTo(models.segment, {
      as: 'segment',
      foreignKey: {
        allowNull: false,
      },
    })

    models.integration.belongsTo(models.user, {
      as: 'createdBy',
    })

    models.integration.belongsTo(models.user, {
      as: 'updatedBy',
    })
  }

  return integration
}
